basic tutorials (#1)
keypom/keypom-docs: docs/Tutorials/Basics/simple-drops.md

---
sidebar_label: 'Simple Drop'
---
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> # Simple Drop

Author: ^nvm

## Prerequisites
For the basic tutorials, you can choose to run the scripts on your own machine. To do son, you must have the following:

1. [Node JS](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm)  
2. [NEAR-API-JS](https://docs.near.org/tools/near-api-js/quick-reference#install)  
3. To install the SDK, simply run the following in your command prompt.  
```bash
npm -i keypom-js
```

---

## Introduction
:::info
In this tutorial, you are going to learn how to create a simple drop from scratch. 

> A simple drop allows you to onboard users onto NEAR and send them any amount of $NEAR. To learn more on what a simple drop is, [click here](../../Concepts/Keypom%20Protocol/Github%20Readme/Types%20of%20Drops/simpledrops.md)

Author: if this is right at the start in the intro, using the term "funding account" can come off a bit confusing. I've changed it to

"Creating a simple drop allows you to embed any amount of $NEAR into a link that can be shared with your friends. When the link is clicked, your friend can be onboarded onto NEAR with a brand new named account. Alternatively, they can send the assets to their existing wallet."

:::

Creating a simple drop is a two step process:  

1) Establishing a connection to the NEAR blockchain.  

> 2) Creating keypairs and the simple drop.  

Author: I've been purposefully leaving out keypair creation in the other tutorials; if we want to explain everything in depth for all basic tutorials, i assume leaving it out is counter intuitive

const { KeyPair, keyStores, connect } = require("near-api-js");
const path = require("path");
const homedir = require("os").homedir();

async function simpleDropKeypom(){
// STEP 1: Initiate a NEAR connection.

// STEP 2: Create keypairs and the drop.
}

simpleDropKeypom()

```

---

## Getting Started - Connecting to NEAR
The first step of creating an Keypom drop is to connect to the NEAR blockchain.

This is done with NEAR-API-JS and consists of creating an unecrypted local keystore and, creating a NEAR configurations, and then using that to initialize a connection to the blockchain. More information about this process can be found [here](https://docs.near.org/tools/near-api-js/quick-reference#key-store).

```js reference
https://github.com/keypom/keypom-js/blob/49244d227f23535ae8962707183d1eca59280d29/docs-examples/keypom-js-sdk/simple-example.js#L8-L25
```

---

## Creating the Simple Drop
Now that the connection to the NEAR blockchain is set up, the next step is to create the Keypairs and then create the drop.  
:::tip
Recall that the Keypairs that are being generated are the ones that will be embedded in the link given to the user. 
:::

### The Keypom SDK Approach

> In the SDK, this process starts with calling the `initKeypom` function. This function will always be the first function you need to call to interact with the SDK. 

Reviewer: This is the first time the user is ever seeing `initKeypom`. We should explain what it does and why it's important / why it's the first function you need to call to interact with the SDK

Reviewer: I think for all the basic tutorials, we should explain things in depth. For advanced tutorials, we can link out to other places in docs.

:::info

> The `initKeypom` function is the first function you need to call in order to interact with the SDK.

Reviewer: This was mentioned 1 line above already

Author: kept the info block one instead, as it appears in all the other tutorials as well

Author: expanded the non-info one. Still keeping the info block one for clarity/skimming

Reviewer: Let's still remove this - info / note tags are generally used to provide more context outside of what is written in the core body. If it's just repeated information, it's kinda strange.

:::

After `initKeypom` is called, the key generation and drop creation can be done. These two tasks are combined into one function, `createDrop`. This function can be either given a set of pre-created keypairs, **or** it can create them automatically.

> ### The NEAR-API-JS Approach

Reviewer: Let's scrap all explanations for NAJ and simply have the extra tab that people can reference to see what exactly is happening behind the scenes with the Keypom SDK

:::tip
Note that with NEAR-API-JS, an attached deposit must be added to the `functionCall`. This is to cover the cost of creating the drop. With simple drops, the cost is the sum of $NEAR embedded in each link as well as the storage costs on the protocol.

In these tutorials, a generous deposit is attached and any unused $NEAR is simply added to the funder's [Keypom balance](../../Concepts/Keypom%20Protocol/balances.md)
:::

<Tabs>
<TabItem value="KPJS" label="🔑Keypom-JS SDK">

```js reference
https://github.com/keypom/keypom-js/blob/49244d227f23535ae8962707183d1eca59280d29/docs-examples/keypom-js-sdk/simple-example.js#L27-L41
```

</TabItem>
<TabItem value="NRJS" label="💻NEAR-API-JS">

```js reference
https://github.com/keypom/keypom-js/blob/49244d227f23535ae8962707183d1eca59280d29/docs-examples/near-api-js/simple-near-example.js#L26-L51
```

</TabItem>
</Tabs>

:::note
All function parameters and default values for the SDK and Keypom functions can be found in the [SDK Typedocs](../../keypom-sdk/modules.md). For NEAR-API-JS functions, their arguments can be found [here](https://docs.near.org/tools/near-api-js/reference)
:::

---

## Full Solution

> The final code, using the skeleton defined in the [introduction](simple-drops.md#introduction) to create a simple 1 $NEAR linkdrop can be seen below.

Reviewer: ```suggestion
Now that everything has been put together, the final code can be seen below.
```

<Tabs>
<TabItem value="KPJS" label="🔑Keypom-JS SDK">

> ```js reference

Reviewer: We need to add a section about generating the linkdrop links cuz this shows up in the final code but isn't mentioned anywhere in the tutorial. We should talk about how for this tutorial, we'll use the NEAR wallet which has a dedicated claim page that expects a URL of the form:

```
wallet.NETWORK.near.org/linkdrop/CONTRACT/PRIVATE_KEY
```

https://github.com/keypom/keypom-js/blob/49244d227f23535ae8962707183d1eca59280d29/docs-examples/keypom-js-sdk/simple-example.js#L1-L55
```

</TabItem>
<TabItem value="NRJS" label="💻NEAR-API-JS">

```js reference
https://github.com/keypom/keypom-js/blob/49244d227f23535ae8962707183d1eca59280d29/docs-examples/near-api-js/simple-near-example.js#L1-L53
```

</TabItem>
</Tabs>

---

## Testing
### Running the Script
Here, you'll learn how to run the code that was just covered, and what to expect.

> To view the completed code, clone the following repo:

Reviewer: ```suggestion
To view the completed code, clone the Keypom SDK repo and visit the examples directory:
```

``` bash
git clone https://github.com/keypom/keypom-js && cd keypom-js/docs-examples/keypom-js-sdk/simple-example.js
```
To run the code you just cloned, install all the necesasry packages. 
```bash
npm install
```

:::caution
Prior to running these scripts, ensure you replace all instances of `keypom-docs-demo.testnet` in the script with the credentials of your account found in your `~/.near-credentials` folder
:::

From there, you are able to run this Simple Drop script that was made in this tutorial using the following command:
``` bash
npm run simple-keypom
```
:::note

> The SDK script is being tested here; use `npm run simple-near` to test the NEAR-API-JS script instead.

Reviewer: Let's remove this since it doesn't make sense to run the NAJ script as it doesn't offer anything different.

Author: if we're offering the NAJ code blocks, shouldn't we also be giving them instructions on how to run them? Otherwise, they'd need to figure out the file layout and manually run it without the shortcut

        Log [v1-3.keypom.testnet]: Total required storage Yocto 11110000000000000000000
        Log [v1-3.keypom.testnet]: Current balance: 1.2083208, 
            Required Deposit: 1.0308726, 
            total_required_storage: 0.01111,
            Drop Fee: 0, 
            Key Fee: 0 Total Key Fee: 0,
            allowance: 0.0187626 total allowance: 0.0187626,
            access key storage: 0.001 total access key storage: 0.001,
            deposits less none FCs: 1 total deposits: 1 lazy registration: false,
            deposits for FCs: 0 total deposits for FCs: 0,
            uses per key: 1
            None FCs: 0,
            length: 1
            GAS to attach: 100000000000000
        Log [v1-3.keypom.testnet]: New user balance 0.1774482
        Log [v1-3.keypom.testnet]: Fees collected 0
Public Keys and Linkdrops:  {
  'ed25519:8nCdZ32nRUdZRXs4dSgjq2FiJnsgMSJYpxNJ4cg7Z1iz': 'https://wallet.testnet.near.org/linkdrop/v1-3.keypom.testnet/5gzHZeLFDyDJHeCVtACwQxbMj8BRTeuu1WVd7eSNhjtc89uZjCoxXW3vh8R1RwEK44w3rb7sBAqMg5hR1MCHsC6e'
}
```

</p>
</details>

### Claiming and Explorer Transactions

> Once you have the link, you are able to claim the linkdrop you've just created. The output link will take you to the following NEAR Wallet page, where you will have the choice to call claim to an existing account or create a new one. 

Reviewer: ```suggestion
Once you have the link, you are able to claim the linkdrop you've just created. Once clicked, it will take you to the following NEAR Wallet page, where you will have the choice to claim with an existing account or create a new one. 
```

<p align="center"> <img src={require("/static/img/docs/basic-tutorials/simple/nw-claim.png").default} alt="Near Wallet claim" width="80%"/> </p>

After the claim transaction succeeds, you can check the transactions on the [NEAR Explorer](https://explorer.near.org/).

> To view the transactions, you can search up the Keypom contract ID: `v1-3.keypom.testnet`. You should be able to see the [`create_drop`](https://explorer.testnet.near.org/transactions/f5HHxMyUELD8oGFpMCQAKcYVhyLAfKyjmXAeFG68FCB#AJ5so8mEDh9XEYNcB7vomTxUDV3BAMyLVA7QMguHuscA) and [`claim`](https://explorer.testnet.near.org/transactions/BApV2YJh8yHUGJ5c4KUnBPNyjAt4znxAFUxPzDXognu5) transactions. 

Reviewer: Let's modify the script to console log the explorer link to the contract. At the end of the script, add:

```
console.log(`Keypom Contract Explorer Link: explorer.${network}.near.org/accounts/${KEYPOM_CONTRACT}.com
```

instead of explaining how to get to this page.

<p align="center"> <img src={require("/static/img/docs/basic-tutorials/simple/explorer.png").default} alt="explorer transactions" width="80%"/> </p>

---

## Conclusion

> In this tutorial, you learned the basic [steps of creating a simple drop](simple-drops.md#introduction), how to [initialize a NEAR blockchain connection](simple-drops.md#getting-started---connecting-to-near), and how to [create the keys and the drop](simple-drops.md#creating-the-simple-drop). You also learned the purpose of the `initKeypom` function and when to use it.

Reviewer: ```suggestion
In this tutorial, you learned the basic steps of [creating a simple drop](simple-drops.md#introduction), how to [initialize](simple-drops.md#getting-started---connecting-to-near) a NEAR blockchain connection, and how to [create](simple-drops.md#creating-the-simple-drop) the keys and drop. You also learned the purpose of the `initKeypom` function and when to use it.
```

---
sidebar_label: 'Simple Drop'
---
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Simple Drop

## Introduction

> In this tutorial, you will learn how to create a Simple $NEAR Drop. This will allow you to onboard users using a simple Web2 style link. Alternatively, if the user has a wallet already, the assets can be sent there instead.

Reviewer: ```suggestion
In this tutorial, you will learn how to create a simple $NEAR drop that allows you to onboard either new or existing users using a simple Web2 style link.
```

> In a simple drop, the assets consist of $NEAR. 

Reviewer: Let's remove. I think it's pretty intuitive that a $NEAR drop contains $NEAR.

<p align="center"> <img src={require("/static/img/docs/basic-tutorials/simple/nw-claim.png").default} alt="Near Wallet claim" width="100%"/> </p>

To learn more about the Simple Drop, see the [concepts page](../../Concepts/Keypom%20Protocol/Github%20Readme/Types%20of%20Drops/simpledrops.md)

---

> ## Prerequisites

Reviewer: I think in this section we should bootstrap the user fully:
1. Create a folder where the project will live.
2. Create a default `package.json` file
3. Create an empty `simple-keypom.js` file.
4. Install keypom SDK using `npm install keypom-js`

This is truly the prerequisites and I think we should add this process in ^

For the basic tutorials, you can choose to run the scripts on your own machine. To do so, you must have the following:

1. [Node JS](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm)  
2. [`NEAR-API-JS`](https://docs.near.org/tools/near-api-js/quick-reference#install)  
3. To [install the SDK](https://github.com/keypom/keypom-js#installation), simply run the following in your command prompt.  
```bash
npm -i keypom-js
```

---

## Breaking Down the Problem

Creating a simple drop is a four step process:  

1) Establishing a connection to the NEAR blockchain.  

> 2) Create a set of access keys where the assets will be stored.  

Reviewer: Let's keep the tense consistent:
```suggestion
2) Creating a set of access keys where the assets will be stored.  
```

> 3) Use the Keypom SDK to create the drop  

Reviewer: ```suggestion
3) Using the Keypom SDK to create the drop  
```

> 4) Create linkdrops  

Reviewer: ```suggestion
4) Creating the linkdrops.
```

The following skeleton code can be used as a starting point:
``` js
// Each of the two methods to create this drop will have their own unique set of imports

// Imports used in the Keypom SDK method:
const { initKeypom, createDrop } = require("keypom-js");
const { KeyPair, keyStores, connect } = require("near-api-js");
const path = require("path");
const homedir = require("os").homedir();

// Imports used in the NEAR-API-JS method:
const { parseNearAmount, formatNearAmount } = require("near-api-js/lib/utils/format");
const { KeyPair, keyStores, connect } = require("near-api-js");
const path = require("path");
const homedir = require("os").homedir();

async function simpleDropKeypom(){
// STEP 1: Initiate a NEAR connection.

// STEP 2: Create a set of access keys 

// STEP 3: Create the drop.

// STEP 4: Create linkdrops
}

simpleDropKeypom()

```

---

## Getting Started - Connecting to NEAR

> The first step of creating an Keypom drop is to connect to the NEAR blockchain.

Reviewer: ```suggestion
The first step to create a Keypom drop is connecting to the NEAR blockchain.
```

This is done with `NEAR-API-JS` and consists of the following steps:

> 1) Create an Keystore, which stores your access keys used to sign transactions   

Reviewer: ```suggestion
1) Create a Keystore, which stores your access keys used to sign transactions   
```

  * select a network, either `testnet` or `mainnet`  

>   * choose a location for the Keystore, either a folder on your local machine, or an in-memory keystore   

Reviewer: ```suggestion
  * choose a location where the keypairs live, either a folder on your local machine, or in-memory   
```

2) Define a NEAR configuration using the Keystore  
3) Use the configuration to initialize a connection to NEAR  

More information about this process can be found [here](https://docs.near.org/tools/near-api-js/quick-reference#key-store).

:::note
For simplicity, this tutorial will choose a file-based keystore and point to the `~/.near-credentials` folder on your local machine since this is where most of your keys are stored. For more information about KeyStores, visit NEAR's [official docs](https://docs.near.org/tools/near-api-js/quick-reference#key-store).
:::

```js reference
https://github.com/keypom/keypom-js/blob/bbe4716ff64dd7a73a6d727a5aea518e8141f60f/docs-examples/keypom-js-sdk/simple-example.js#L8-L25
```

---

## Creating the Simple Drop
Now that the connection to the NEAR blockchain is set up, the next step is to create the Keypairs and the drop using the Keypom SDK.

In the SDK, this process starts with calling the `initKeypom` function. This function will always be the first function you need to call to interact with the SDK. 

> `initKeypom` initializes the SDK to allow for interactions with the Keypom Protocol. Without it, none of the other SDK functions would work as expected. If a NEAR connection is not already present,it will initialize a new one for you. More info on the `initKeypom` function can be found [here](../../keypom-sdk/modules#initkeypom).

Reviewer: ```suggestion
`initKeypom` initializes the SDK to allow for interactions with the Keypom smart contracts. Without it, none of the other SDK functions would work as expected. If a NEAR connection is not already present, it will initialize a new one for you. More info on the `initKeypom` function can be found [here](../../keypom-sdk/modules#initkeypom).
```
